src/Cli/dotnet/commands/dotnet-run/VirtualProjectBuildingCommand.cs

```diff
@@ -139,86 +178,443 @@ private ProjectInstance CreateProjectInstance(
+                    """);
+            }
+            else
+            {
+                writer.WriteLine($"""
+                      <Sdk Name="{EscapeValue(sdk.Name)}" Version="{EscapeValue(sdk.Version)}" />
+                    """);
+            }
+
+            processedDirectives++;
+        }
+
+        if (processedDirectives > 1)
+        {
+            writer.WriteLine();
+        }
+
+        // Kept in sync with the default `dotnet new console` project file (enforced by `DotnetProjectAddTests.SameAsTemplate`).
+        writer.WriteLine($"""
               <PropertyGroup>
-            {CommonProjectProperties}
+                <OutputType>Exe</OutputType>
+                <TargetFramework>net10.0</TargetFramework>
+                <ImplicitUsings>enable</ImplicitUsings>
+                <Nullable>enable</Nullable>
               </PropertyGroup>
+            """);
+
+        if (virtualProjectFile)
+        {
+            writer.WriteLine("""
+
+                  <PropertyGroup>
+                    <EnableDefaultItems>false</EnableDefaultItems>
+                  </PropertyGroup>
+                """);
+        }
+
+        if (propertyDirectives.Any())
+        {
+            writer.WriteLine("""
+
+                  <PropertyGroup>
+                """);
+
+            foreach (var property in propertyDirectives)
+            {
+                writer.WriteLine($"""
+                        <{property.Name}>{EscapeValue(property.Value)}</{property.Name}>
+                    """);
+
+                processedDirectives++;
+            }
+
+            writer.WriteLine("  </PropertyGroup>");
+        }
+
+        if (packageDirectives.Any())
+        {
+            writer.WriteLine("""
+
+                  <ItemGroup>
+                """);
+
+            foreach (var package in packageDirectives)
+            {
+                if (package.Version is null)
+                {
+                    writer.WriteLine($"""
+                            <PackageReference Include="{EscapeValue(package.Name)}" />
+                        """);
+                }
+                else
+                {
+                    writer.WriteLine($"""
+                            <PackageReference Include="{EscapeValue(package.Name)}" Version="{EscapeValue(package.Version)}" />
+                        """);
+                }
+
+                processedDirectives++;
+            }
+
+            writer.WriteLine("  </ItemGroup>");
+        }
+
+        Debug.Assert(processedDirectives + directives.OfType<CSharpDirective.Shebang>().Count() == directives.Length);
+
+        if (virtualProjectFile)
+        {
+            Debug.Assert(targetFilePath is not null);
+
+            writer.WriteLine($"""
+
+                  <ItemGroup>
+                    <Compile Include="{EscapeValue(targetFilePath)}" />
+                  </ItemGroup>
+
+                """);
+
+            foreach (var sdk in sdkDirectives)
```

Comment:
You don't have to Skip(1)? And perhaps this is nitpicky, but shouldn't you go in reverse order?

Reply:
> You don't have to Skip(1)?

No, that's only needed when the first Sdk is put into `<Project Sdk>` attribute but that never happens for virtual project files. Instead they get all sdks emitted as `<Import Project="Sdk.props" />` elements at the beginning and `<Import Project="Sdk.targets" />` at the end.

> shouldn't you go in reverse order?

I considered that but as far as I can tell, the following

```csproj
<Project>
<Sdk Name="1" />
<Sdk Name="2" />
</Project>
```

is equivalent to

```csproj
<Project>
<Import Project="Sdk.props" Sdk="1" />
<Import Project="Sdk.props" Sdk="2" />
<Import Project="Sdk.targets" Sdk="1" />
<Import Project="Sdk.targets" Sdk="2" />
</Project>
```

Reply:
here's the MSBuild logic for these implicit imports: https://github.com/dotnet/msbuild/blob/f1b335d0aa9423c304b5cb186bad40935a6e2f2b/src/Build/Construction/ProjectRootElement.cs#L1903-L1932

followed by https://github.com/dotnet/msbuild/blob/f1b335d0aa9423c304b5cb186bad40935a6e2f2b/src/Build/Definition/Project.cs#L2860-L2883 for the actual loading.

Short version: @jjonescz is correct here. it's not a nesting-doll model, it's an in-order Import for both the Sdk.props and Sdk.targets entrypoints.